File: rust/src/commands/diff.rs
use crate::commands::shared::{normalize_env_label, report_parse_errors};
use crate::core::pipeline::load_project;
use crate::detectors::environment_diff::compare_environments;
use camino::Utf8PathBuf;
use clap::Args;
use std::collections::BTreeSet;

#[derive(Args, Debug)]
pub struct DiffArgs {
    /// Project root (default: current directory)
    #[arg(long, short = 'C')]
    pub root: Option<Utf8PathBuf>,

    /// First environment (e.g. development, production, or dev/prod aliases)
    pub env_a: String,

    /// Second environment
    pub env_b: String,

    /// Print the diff as JSON
    #[arg(long)]
    pub json: bool,
}

/// `envdoctor diff <env1> <env2>` — compare variable sets across environments.
pub async fn diff(args: DiffArgs) -> Result<u8, anyhow::Error> {
    let root = args.root.unwrap_or_else(|| Utf8PathBuf::from("."));
    let root: Utf8PathBuf = std::fs::canonicalize(&root)?
        .try_into()
        .map_err(|_| anyhow::anyhow!("Invalid root path"))?;

    let (model, _config) = load_project(&root).await;
    let label_a = normalize_env_label(&args.env_a);
    let label_b = normalize_env_label(&args.env_b);

    let available: BTreeSet<String> = model
        .env_files
        .iter()
        .filter_map(|f| f.environment.clone())
        .filter(|e| e != "example")
        .collect();

    report_parse_errors(&model, &root);

    if !available.contains(&label_a) || !available.contains(&label_b) {
        if !available.contains(&label_a) {
            eprintln!("error Environment \"{label_a}\" has no files in this project.");
        }
        if !available.contains(&label_b) {
            eprintln!("error Environment \"{label_b}\" has no files in this project.");
        }
        let list: Vec<&String> = available.iter().collect();
        let joined = if list.is_empty() {
            "none".to_string()
        } else {
            list.iter().map(|s| s.as_str()).collect::<Vec<_>>().join(", ")
        };
        eprintln!("  Available: {joined}");
        return Ok(2);
    }

    let entries = compare_environments(&model, &label_a, &label_b);
    let missing_count = entries.iter().filter(|e| !e.present_in_both).count();

    if args.json {
        let variables: Vec<serde_json::Value> = entries
            .iter()
            .map(|e| {
                let missing_in = if e.present_in_both {
                    serde_json::Value::Null
                } else if e.present_in_a {
                    serde_json::json!(label_b)
                } else {
                    serde_json::json!(label_a)
                };
                serde_json::json!({
                    "name": e.name,
                    "status": if e.present_in_both { "same" } else { "missing" },
                    "missingIn": missing_in,
                })
            })
            .collect();
        let out = serde_json::json!({
            "environments": [label_a, label_b],
            "exitCode": if missing_count > 0 { 1 } else { 0 },
            "total": entries.len(),
            "missing": missing_count,
            "variables": variables,
        });
        println!("{}", serde_json::to_string_pretty(&out)?);
        return Ok(if missing_count > 0 { 1 } else { 0 });
    }

    println!("ENVIRONMENT DIFF");
    println!("{}\n", "=".repeat("ENVIRONMENT DIFF".len() * 2));
    println!("  {label_a} → {label_b}\n");

    for entry in &entries {
        if entry.present_in_both {
            println!("  = {}  present in both", entry.name);
        } else if entry.present_in_a {
            println!("  ✗ {}  missing in {}", entry.name, label_b);
        } else {
            println!("  ✗ {}  missing in {}", entry.name, label_a);
        }
    }
    println!(
        "\n  Summary: {} variables · {} missing · {} present in both",
        entries.len(),
        missing_count,
        entries.len() - missing_count
    );

    Ok(if missing_count > 0 { 1 } else { 0 })
}

File: dart/lib/src/commands/diff.dart
/// `envdoctor diff <env1> <env2>` — compare variable sets across environments.
library;

import '../core/pipeline.dart';
import '../detectors/environment_diff.dart';
import '../utils/json.dart';
import 'shared.dart';

class DiffOptions {
  final String rootDir;
  final String envA;
  final String envB;
  final bool json;

  DiffOptions({
    required this.rootDir,
    required this.envA,
    required this.envB,
    this.json = false,
  });
}

int runDiff(DiffOptions opts) {
  final context = loadProject(opts.rootDir);
  final labelA = normalizeEnvLabel(opts.envA);
  final labelB = normalizeEnvLabel(opts.envB);

  final available = <String>{};
  for (final f in context.model.envFiles) {
    if (f.environment != null && f.environment != 'example') {
      available.add(f.environment!);
    }
  }

  reportParseErrors(context.model, opts.rootDir);

  if (!available.contains(labelA) || !available.contains(labelB)) {
    if (!available.contains(labelA)) {
      err('error Environment "$labelA" has no files in this project.\n');
    }
    if (!available.contains(labelB)) {
      err('error Environment "$labelB" has no files in this project.\n');
    }
    err('  Available: ${available.isEmpty ? 'none' : available.join(', ')}\n');
    return 2;
  }

  final entries = EnvironmentDiffDetector.compareEnvironments(
      context.model, labelA, labelB);
  final missingCount = entries.where((e) => !e.presentInBoth).length;

  if (opts.json) {
    final variables = entries
        .map((e) => JsonObject()
          ..add('name', e.name)
          ..add('status', e.presentInBoth ? 'same' : 'missing')
          ..add('missingIn',
              e.presentInBoth ? null : (e.presentInA ? labelB : labelA)))
        .toList();
    final payload = JsonObject()
      ..add('environments', [labelA, labelB])
      ..add('exitCode', missingCount > 0 ? 1 : 0)
      ..add('total', entries.length)
      ..add('missing', missingCount)
      ..add('variables', variables);
    out('${jsonPretty(payload)}\n');
    return missingCount > 0 ? 1 : 0;
  }

  out('ENVIRONMENT DIFF\n');
  out('${'─' * ('ENVIRONMENT DIFF'.length * 2)}\n\n');
  out('  $labelA → $labelB\n\n');

  for (final entry in entries) {
    if (entry.presentInBoth) {
      out('  ✓ ${entry.name}  present in both\n');
    } else if (entry.presentInA) {
      out('  ❌ ${entry.name}  missing in $labelB\n');
    } else {
      out('  ❌ ${entry.name}  missing in $labelA\n');
    }
  }
  out(
      '\n  Summary: ${entries.length} variables · $missingCount missing · ${entries.length - missingCount} present in both\n');

  return missingCount > 0 ? 1 : 0;
}
